Publish versioned docs

ffakenz opened the pull request:
## Summary
Publish docs from last release + unstable separately onto hydra.family

## Why

We want both, the last released, but also the bleedging edge documentation on our website.
Users should see by default the last released version, but we as developer want to have a view on the current `master` version of the docs.

## What

* The standard paths of the website should be the one of the last released (= last tagged) version, we call it `latest`
  - Nice to have: Show the last released version in the top, e.g. 0.9.0 
* We also publish the documentation for `master` version of the repository, we call it `unstable`

* The monthly report from `master` (`unstable`) should also be the one included in the `latest`
 
* Accessing the versions through the URL only is okay
  * Nice to have: switch between version `unstable` and the released versions

Out of scope:
* No feature branch or PR versions of the documentation

## How

* Keep it simple
> The idea is to generate doc builds using different base URLs in config, then merge them into a single one, and finally publish it to gh-pages as a whole, having a main index that redirects to the latest release version. 
This is the current implementation.

* Github environments?
> Not explored yet.

* Docusaurus versions?
> Mithril is using this feature ([reference](https://github.com/input-output-hk/mithril/commit/f2558c5f6acf47a6b0f8bce1bde3e5a9def6b3aa)) but it requires us to maintain a versioned folder in docs and that is something we want to avoid.
So this effort was explored but discontinued.

--

* [x] CHANGELOG updated
* [x] Documentation updated
* [x] Added and/or updated haddocks
* [x] No new TODOs introduced or explained herafter

ffakenz committed:
Fix global links to be relative.

This will help later when versioning docs,
because different versions will have different base url.

ffakenz committed:
Update CI to upload unstable docs.

Next step will be responsible for merging released and unstable docs.

ffakenz committed:
Add workflow to produce versioned docs.

This will produce a single docs build combining both released and
unstable.

This workflow will only get triggered if CI completed with success.

Currently it is only uploading the artifact result.
Soon it will publish docs to gh-pages.

ffakenz committed:
Fix links in glossary: avoid using relative paths as they are discourage by docusaurus.

ffakenz committed:
Fix broken haddock links to make yarn build succeed.

ffakenz committed:
Fix broken haddock links by using reference paths and avoid throwing on onBrokenLinks but mark them as warn

ffakenz committed:
Alter onBrokenLinks config to throw.

Removed link-checker workflow as we see no value on getting a report
when it is too late, as the web is already live.

Beware: for docusaurus build to be happy we need to
generate /haddock and /benchmark docs as done in workflow ci.

v0d1ch commented on docs/benchmarks/profiling.md:
```diff
@@ -10,7 +10,7 @@ contributors to the `hydra-node`.
 costs in size, memory and cpu usage of the Hydra protocol transactions on
 Cardano. The latest results can be seen
-[here](https://hydra.family/head-protocol/benchmarks/transaction-cost/).
+[here](/benchmarks/transaction-cost/).
```
Is there a reason to change to absolute url? For me both are fine just curious.

ffakenz replied:
It is a good question, and [here](https://docusaurus.io/docs/versioning#link-docs-by-file-paths) is the reason behind it.

v0d1ch commented on .github/workflows/docs.yaml:
```diff
@@ -0,0 +1,263 @@
+    #     publish_dir: docs/public
+    #     enable_jekyll: true
+    #     force_orphan: true
+    
```
Just nuke these lines if we don't need them.

ffakenz replied:
we need them.

once the PR gets approved I need to Publish Documentation.

now my question is: should we keep the published artifact?

ffakenz replied:
We agreed during tactical to keep it.

v0d1ch reviewed (changes requested):
Looks good, if it works :ship: but let's remove leftover TODO

ffakenz committed:
Remove schedule for ci workflow

ffakenz committed:
Add two new workflows to take care of:
- release process (generate release docs + publish to site)
- publish new unstable (master) versions to site

previos  workflow was removed in favor of the above.

the release workflow will only get triggered when a tag is pushed,
whereas the ci-publish will run once ci its successfully merged to master.

pgrange commented on .github/workflows/ci-publish.yaml:
```diff
@@ -0,0 +1,47 @@
+    runs-on: ubuntu-latest
+    if: ${{ github.event.workflow_run.conclusion == 'success' }}
+    steps:
+    - name: 📥 Download stable docs with released
+      uses: dawidd6/action-download-artifact@v2
+      with:
+        workflow: release.yaml
+        workflow_conclusion: success
+        name: stable
+        path: docs/public/
```
Question: how can you be sure from which workflow run your fetching artifacts? I didn't find any documentation in the GitHub action as to which policy it uses to select between several run. Did you find something about it?

ffakenz replied:
Good question!

On [the action repo](https://github.com/dawidd6/action-download-artifact) at the very end, there is an option that says:

```yaml
# Optional, check the workflow run to whether it has an artifact
# then will get the last available artifact from the previous workflow
# default false, just try to download from the last one
check_artifacts:  false
```
So from my understanding, it seems to be fetching the latest always, and from my empiric manual tests, that is the behavior I observe.
I can tell from [the code](https://github.com/dawidd6/action-download-artifact/blob/master/main.js#L122) that it tries to fetch the artifact from a job using its same run-id as the triggering workflow run-id, and in case there is none, it finds the first occurrence of the artifact among [all the existing ones](https://github.com/dawidd6/action-download-artifact/blob/master/main.js#L163) in [descending order](https://github.com/dawidd6/action-download-artifact/blob/master/main.js#L105).

Hope I am not confused.
Let me know your thoughs.

pgrange commented on .github/workflows/ci-publish.yaml:
```diff
@@ -0,0 +1,47 @@
+    #     github_token: ${{ secrets.GITHUB_TOKEN || github.token }}
+    #     publish_dir: docs/public
+    #     enable_jekyll: true
+    #     force_orphan: true
```
I would uncomment the following lines

ffakenz replied:
they must be uncommented before merging to master :)

pgrange commented on .github/workflows/ci-publish.yaml:
```diff
@@ -0,0 +1,47 @@
+        name: unstable
+        path: docs/public/unstable
+
+    - name: 💾 Upload artifact
+      uses: actions/upload-artifact@v3
+      with:
+        name: head-protocol
+        path: docs/public
```
I would remove this as it is my understanding that it only re-upload some artifact it downloaded before. If it is needed, I would add an explanation of why we need to do that.

ffakenz replied:
this artifact is the result of merging the latest release with master.

we agreed during tactical these artifacts were valuable for debugging purposes.

do you think they are not needed anymore?

pgrange replied:
I don't think they're valuable anymore but that's your call 👍

ffakenz replied:
We will keep it for the moment and maybe remove it in the near future in case of need.

pgrange reviewed (approved):
This review tries to apply the [perfection game protocol](https://liveingreatness.com/core-protocols/perfection-game/)
 I will describe what I like about this P.R. and then what improvement I can think of to make it perfect. Ask for clarifications if my comments are not clear.
 If you disagree with my comments, just ignore them and don’t answer or explain. Hence I, in advance, approve the P.R.
 

What I like about this P.R.:
* Solves the problem, appreciate all the work you've put into this
* Some code duplication that can be solved later but which make sense at this point in time (we want to see it work for real at least once)
* the fact that you practiced on a fork of the project

For me to find it perfect you would have to:
* I would clean the git history, to make the review easier, as I see some back and forth in the changes (some examples below)
* I would remove commit [Add workflow to produce versioned docs](https://github.com/input-output-hk/hydra/pull/783/commits/14f349e7978289bd9bd39e3ac53d2e074c666402) from the git history as it is my understanding that these changes are not present in the end
* I would amend commit [Fix broken haddock](https://github.com/input-output-hk/hydra/pull/783/commits/f259b27492d6f5cc85710dd34c17c03140b2b1bc)  to remove the change in docusaurus.config.js as this has disappeared in the final version
* I would of course amend commit [Alter onBrokenLinks config to throw](https://github.com/input-output-hk/hydra/pull/783/commits/fc4ae924289c87eb089802775e64cf76420c7113) to not mention the warn/throw removed from the previous commit
* I would amend commit [Remove schedule for ci workflow](https://github.com/input-output-hk/hydra/pull/783/commits/93e16a18a5883a3e4012bd1fde5b762f78dd52a3) to explain why this schedule is removed or, I would remove this commit and make it part of another P.R. if it is not necessary here.
* see other improvement proposals inline

ffakenz committed:
Enable job to publish the built documentation to our web site

ffakenz commented:
Visual notes on how the new workflows outcomes look like

<img width="971" alt="image" src="https://user-images.githubusercontent.com/14299604/228801770-59712dd4-4b85-4157-a07f-a49d3f899b76.png">

<img width="622" alt="image" src="https://user-images.githubusercontent.com/14299604/228801884-a26aa9f3-aa59-415f-a3cb-207deefc202c.png">

And this is how the **Publish Documentation** steps look like

<img width="449" alt="image" src="https://user-images.githubusercontent.com/14299604/228802224-cf1e5dcc-0c6e-4ba0-acd0-a4c4d5eaf16a.png">

Finally, the CI workflow executes the same steps but it uploads the unstable version of docs instead of publishing them to the website. This unstable artifact is later fetched by both ci-publish and release workflows to generate the final site to be published.
